### ISSUE TITLE
Secure Secrets for Generalized Continuous Deploys

### ISSUE THREAD

**Issue by 0x4007** [ubiquity/cloudflare-deploy-action#1]
### Context

I spent quite a bit of time cleaning up the code and trying to make the deploys more general purpose so that we can include in our `@ubiquity/ts-template` to make it seamless for contributors to open pull requests and for our reviewers to be able to review their work. I severely underestimated the amount of time that this would take. However:

- I finally got it stable and fairly fast (I removed unnecessary steps) over the past few days. 
- Here's a test on a pull request: https://github.com/ubiquity/work.ubq.fi/pull/16#issuecomment-1962506213. 
   - I also fixed the details around commit hashes posting in chronological order, which used to be an issue. 
- Specifically the work is done in [build.yml](https://github.com/ubiquity/work.ubq.fi/blob/development/.github/workflows/build.yml) and then its passed along to this current repository (`@ubiquity/cloudflare-deploy-action`) starting with [action.yml](https://github.com/ubiquity/cloudflare-deploy-action/blob/06783ab3b6bf9e91f82ae5be48e5ac18def8b9fc/action.yml) in the root.
- From there, it goes to one script for dealing with cloudflare: [cloudflare-deploy.sh](https://github.com/ubiquity/cloudflare-deploy-action/blob/06783ab3b6bf9e91f82ae5be48e5ac18def8b9fc/.github/cloudflare-deploy.sh)
- And then finally it passes the output (the deployed URL) to the node script in [src/index.ts](https://github.com/ubiquity/cloudflare-deploy-action/blob/06783ab3b6bf9e91f82ae5be48e5ac18def8b9fc/src/index.ts) which posts the comment as the bot to the pull request (or commit - but this is commented out now.)

As you can see I was struggling with passing secrets around. I decided to review the old implementation and realized that the clever workaround to expose the secrets was to invoke the action "internally" from the successful completion of a previous action[^1^], as seen [here](https://github.com/ubiquity/pay.ubq.fi/blob/a57f08925a2fa10c756d036b0e5aeb7b54cf81b0/.github/workflows/continuous-deploy.yml#L5-L7). 

I spent a lot more time than I anticipated on this project, and unfortunately I'm starting to run out of development time leading up to 1 March and I still have several other matters I need to tend to. Unfortunately I did not follow best practices and exposed all of what should be secrets as plaintext and hardcoded in the scripts. 

### Notes

- I created a dedicated Cloudflare account specifically for our automatic deployments across all Ubiquity repositories. The API key that is used is tightly scoped just to create and edit Workers deployments. This isn't really a security concern because my updated deploy logic will only deploy to `production` if its merged to the default branch (for most of our repositories its `development` but sometimes `main` - its simple to change in the repository settings) otherwise it will just make a preview on a unique cloudflare URL. 
- I created a dedicated GitHub App (Ubiquibot Continuous Deploys) with minimal permissions to just be able to add comments on pull requests and commits for posting the deploys. 

I don't think its fatal that these credentials are leaked, but it obviously is not following best practices. 

### Objectives

1. Update `@ubiquity/ts-template` with the correct action initiators (`init.yml`, then `build.yml` so that it can have access to secrets)
   - I think `init.yml` would be a dummy action; or to be smart we could make it handle `conventional-commits` and then run the build. All secrets should be available in `build.yml` so that it can apply our i.e. `SUPABASE_ANON_KEY` and `SUPABASE_URL` in the build. That way the reviewers can see the built (and functional) app easily. 
2. Update `@ubiquity/cloudflare-deploy-action` to handle the event coming in from `build.yml`. Now it should be privleged and be able to read our secrets normally. Change my code to use the secrets instead of the hardcoded plaintext values. 
3. Rotate the credentials. 

### Reference

Here is the order of the old setup. 

1. https://github.com/ubiquity/pay.ubq.fi/blob/ca90b265210fe953d942387791a52bca879b538c/.github/workflows/build.yml
2. https://github.com/ubiquity/pay.ubq.fi/blob/a57f08925a2fa10c756d036b0e5aeb7b54cf81b0/.github/workflows/continuous-deploy.yml
3. https://github.com/ubiquity/.github/blob/main/.github/workflows/deploy.yml

### Final Remarks

I was able to make it stable without using download/upload artifact; and without using `checkout` actions. We should try not using these because if I recall correctly, they were the biggest bottlenecks for the slowdowns in the runs. Right now the biggest bottleneck that I couldn't' really get around is the `yarn install` for running node. I think its fine to leave in. 

There was some bug with the wrangler CLI that I couldn't verify if the project already existed remotely. I must have been doing something wrong but I spent so many hours on it that I just decided to [replace it with a curl](https://github.com/ubiquity/cloudflare-deploy-action/blob/06783ab3b6bf9e91f82ae5be48e5ac18def8b9fc/.github/cloudflare-deploy.sh#L28-L31).

[^1^]: As explained plainly by ChatGPT: https://chat.openai.com/share/0455a2fe-1dd4-483b-86e5-b05e025ad485

**Comment by 0x4007** [comment]
@rndquu please spend some time on this so that we can standardize deploys across all of our future projects via `ts-template`. I would appreciate if you can address this within the next couple of weeks.

**Comment by 0x4007** [comment]
Related: https://github.com/ubiquity/ts-template/issues/10

**Comment by rndquu** [comment]
> I created a dedicated Cloudflare account specifically for our automatic deployments across all Ubiquity repositories.

I have access to the `Ubiquity DAO Workers` cloudflare account. Is it the one you're talking about?

> I created a dedicated GitHub App (Ubiquibot Continuous Deploys) with minimal permissions to just be able to add comments on pull requests and commits for posting the deploys.

Can't find it here https://github.com/marketplace?category=&type=&verification=&query=Ubiquibot+Continuous+Deploys

**Comment by 0x4007** [comment]
> > I created a dedicated Cloudflare account specifically for our automatic deployments across all Ubiquity repositories.
> 
> 
> 
> I have access to the `Ubiquity DAO Workers` cloudflare account. Is it the one you're talking about?


Yes


> 
> 
> 
> > I created a dedicated GitHub App (Ubiquibot Continuous Deploys) with minimal permissions to just be able to add comments on pull requests and commits for posting the deploys.
> 
> 
> 
> Can't find it here https://github.com/marketplace?category=&type=&verification=&query=Ubiquibot+Continuous+Deploys

It's not public. It's the one posting all the continuous deploys now. It's called ubiquibot-continuous-deploys

**Comment by 0x4007** [comment]
Hey just following up here. 

I made some special logic for if we create a new repository named *.ubq.fi, leveraging our continuous deploys, it will automatically map the subdomain live on to *.ubq.fi

It would be nice for the keys to be secured or else there will be some manual cleanup required in the `Ubiquity DAO Workers` account for any pollution of the namespace.

**Comment by rndquu** [comment]
> Hey just following up here.
> 
> I made some special logic for if we create a new repository named *.ubq.fi, leveraging our continuous deploys, it will automatically map the subdomain live on to *.ubq.fi
> 
> It would be nice for the keys to be secured or else there will be some manual cleanup required in the `Ubiquity DAO Workers` account for any pollution of the namespace.

Hey, I remember about this task, I'll check it this week

**Comment by 0x4007** [comment]
I just realized that these will deploy to our account even from forks that are not related to Ubiquity. https://github.com/pavlovcik/ubiquity-dollar/actions/runs/8129516826/job/22216713105

I added the template CI in my fork of the Dollar and it ended up on Cloudflare. In the specified process, when its a fork opening up to our organization that should naturally fix this issue.

<img width="929" alt="image" src="https://github.com/ubiquity/cloudflare-deploy-action/assets/4975670/1c40a093-46d9-4352-9c9d-aa24e3b3b79d">

**Comment by rndquu** [comment]
> I just realized that these will deploy to our account even from forks that are not related to Ubiquity. https://github.com/pavlovcik/ubiquity-dollar/actions/runs/8129516826/job/22216713105
> 
> I added the template CI in my fork of the Dollar and it ended up on Cloudflare. In the specified process, when its a fork opening up to our organization that should naturally fix this issue.
> 
> <img alt="image" width="929" src="https://private-user-images.githubusercontent.com/4975670/309546533-1c40a093-46d9-4352-9c9d-aa24e3b3b79d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk1NDA0NDcsIm5iZiI6MTcwOTU0MDE0NywicGF0aCI6Ii80OTc1NjcwLzMwOTU0NjUzMy0xYzQwYTA5My00NmQ5LTQzNTItOWM5ZC1hYTI0ZTNiM2I3OWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDRUMDgxNTQ3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODQwY2E4ZDA0MjU3ZTE2ZjQxY2JhMGY1NjIyNmZmMzM2ODYyM2I2OWMxOTQyNGYwZjAyZWVhMTRiNmI4MmFiOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.De-J60PGiiampj0kXBiIxJ4-V5yxB-3DHiOyx2SUTU8">

So what's the expected behaviour? When PR is from a fork then we deploy to our own account, right?

**Comment by 0x4007** [comment]
This is intended to facilitate our reviewers by posting the deployments on the pull request. So it should only run in the context of a pull request opened against the Ubiquity organization's repositories.

**Comment by 0x4007** [comment]
Is there a problem? Why did you reopen?

**Comment by rndquu** [comment]
> Is there a problem? Why did you reopen?

I will:
1. Pass cloudflare API token and account id [here](https://github.com/ubiquity/ts-template/blob/397bf995a43b8af69b6dfd01a965b8f2f8b2b6b0/.github/workflows/build.yml#L35)
2. Rotate cloudflare API token in github secrets

**Comment by 0x4007** [comment]
I suppose it would also make sense to update the active repo actions as well?

**Comment by rndquu** [comment]
> I suppose it would also make sense to update the active repo actions as well?

yes

**Comment by rndquu** [comment]
@pavlovcik 

TLDR; I want to revert all of the changes ([one](https://github.com/ubiquity/ts-template/pull/32), [two](https://github.com/ubiquity/ts-template/pull/35), [three](https://github.com/ubiquity/onboard.ubq.fi/pull/15)) and keep your initial approach with hardcoding cloudflare API credentials for now

Long story

Github CI has 2 contexts:
1. Target org/repo (for example https://github.com/ubiquity organization)
2. Base repo (where PR is originated)

Github secrets are available in workflows triggered by PRs in the following cases:
1. PR is opened in a branch within the target organization (i.e. contributor is a member of an organization hence he is trusted to create branches)
2. PR is opened in a branch from a forked repository. By default such PRs don't have access to organization secrets but there are options:
  a) Use [pull_request_target](https://github.com/ubiquity/test-cf-deploy/actions/runs/8295132088/workflow#L4)
  b) Use [workflow_run](https://github.com/ubiquity/test-cf-deploy/actions/runs/8295450313/workflow#L4)
  c) Use [pull_request_target](https://github.com/ubiquity/test-cf-deploy/actions/runs/8295132088/workflow#L4) + "deploy" label

All of the options above are a [security risk](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/) (together with the [checkout action](https://github.com/ubiquity/test-cf-deploy/actions/runs/8295450313/workflow#L17-L20)) since they run untrusted code in a PR with access to org/repo secrets and github API token with `write` access.

I've tried it myself and was able to exfiltrate env variables:
1. [Example](https://github.com/ubiquity/test-cf-deploy/actions/runs/8295132088/job/22701554770) with `pull_request_target` (check the "Cypress run" step, you may find `GITHUB_TOKEN` there)
2. [Example](https://github.com/ubiquity/test-cf-deploy/actions/runs/8295450313/job/22702509377) with `workflow_run` (check the "Build" step)

It doesn't really matter if we use a github composite action (like https://github.com/ubiquity/cloudflare-deploy-action) or a reusable workflow:
- If we use a github composite action we can only pass secrets in action input params since github composite actions [don't have access](https://github.com/orgs/community/discussions/51280#discussioncomment-8726096) to org/repo secrets
- If we use a [reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows) then we can access secrets there only if the caller has access to secrets ([Docs](https://docs.github.com/en/actions/using-workflows/reusing-workflows), TLDR; `When a reusable workflow is triggered by a caller workflow, the github context is always associated with the caller workflow.`). It means that we must use either `pull_request_target` either `workflow_run` which is again a security risk.

Overall this issue (unable to run deployment preview from forked repos) is common:
- https://github.com/cloudflare/pages-action/issues/94
- https://github.com/rossjrw/pr-preview-action/issues/54

Our main goal is to make contributors UX as seamless as possible. We could:
1. Create 2 cloudflare accounts: one for deployment previews and another one for production builds (like https://pay.ubq.fi/, https://work.ubq.fi/, etc...)
2. Update "build and deploy" workflow this way (rough code):
```yml
name: Build & Deploy

on:
  push:
    branches:
      - development
  pull_request:

  - name: Deploy to Cloudflare
    uses: ubiquity/cloudflare-deploy-action@main
    with:
      repository: ${{ github.repository }}
      production_branch: ${{ github.event.repository.default_branch }}
      output_directory: "static"
      current_branch: ${{ github.ref_name }}
      pull_request_number: ${{ github.event.pull_request.number }}
      cloudflare_account_id: ${{ secrets.CLOUDFLARE_ACCOUNT_ID }}
      cloudflare_api_token: ${{ secrets.CLOUDFLARE_API_TOKEN }}
      commit_sha: ${{ github.event.pull_request.head.sha }}
```
The main idea is that on the `push` event (to production branch) secrets are available and will be passed while on `pull_request` event secrets will be empty.
3. Update https://github.com/ubiquity/cloudflare-deploy-action so that:
- if cloudflare credentials are passed then use them (which means production deploy)
- if cloudflare credentials are not passed then use hardcoded values for deployment preview

So with the described approach:
- we use hardcoded cloudflare credentials for deployment previews
- we use github secrets to store cloudflare credentials for production builds
- we don't have access to org/repo secrets in the [build step](https://github.com/ubiquity/pay.ubq.fi/blob/da14f1b58eb7d3a1754706bc0c472190c22db968/.github/workflows/build.yml) because it requires using `pull_request_target/workflow_run` + running untrusted code in a PR with access to org/repo secrets

**Comment by rndquu** [comment]
Or use the approach we used earlier:
1. [Build workflow](https://github.com/ubiquity/pay.ubq.fi/blob/a57f08925a2fa10c756d036b0e5aeb7b54cf81b0/.github/workflows/build.yml) creates static files to publish and saves to github artifacts (here the "build workflow" runs in an unprivileged context so we'll have to hardcode env variables required for build)
2. [Cloudflare deploy](https://github.com/ubiquity/pay.ubq.fi/blob/a57f08925a2fa10c756d036b0e5aeb7b54cf81b0/.github/workflows/continuous-deploy.yml) workflow deploys the artifact (this workflow runs in privileged context with access to github secrets so we don't need to maintain 2 accounts)

**Comment by 0x4007** [comment]
Well currently the domain is managed on my private account, and has a worker thats forwarding the subdomains to other workers on the "Ubiquity DAO Workers" account. 

The other "Ubiquity DAO Workers" account only deploys workers that are forwarded to. So because of this, especially during our current rapid development phase, our security downsides should be fairly limited (we don't really have end users using our UIs now.) 

The worst case scenario is that a bad pull is merged and one of our production UIs are compromised. But aside from that, I am unsure that the keys with the current minimum permissions are enough to cause real harm on the dedicated deploys account. It might be worth an audit from you.
